File: cpp/03_advanced/src/PowerUp.cpp
#include "../include/PowerUp.h"
#include <QRandomGenerator>

PowerUp::PowerUp() : duration(0), lifetime(300), age(0) {}

void PowerUp::spawn(const QVector<QPoint>& occupiedPositions) {
    int maxAttempts = 100;
    
    do {
        position.rx() = QRandomGenerator::global()->bounded(20);
        position.ry() = QRandomGenerator::global()->bounded(20);
    } while (occupiedPositions.contains(position) && maxAttempts-- > 0);
    
    // Случайный тип power-up
    int r = QRandomGenerator::global()->bounded(5);
    type = static_cast<PowerUpType>(r);
    
    switch (type) {
        case PowerUpType::SPEED_BOOST:
            color = QColor(0, 255, 255);
            duration = 500;
            break;
        case PowerUpType::INVINCIBILITY:
            color = QColor(255, 255, 0);
            duration = 500;
            break;
        case PowerUpType::SLOW_TIME:
            color = QColor(255, 0, 255);
            duration = 500;
            break;
        case PowerUpType::EXTRA_LIFE:
            color = QColor(255, 100, 100);
            duration = 0;
            break;
        case PowerUpType::SCORE_MULTIPLIER:
            color = QColor(255, 200, 0);
            duration = 0;
            break;
    }
    
    age = 0;
}

void PowerUp::update() {
    age++;
}

bool PowerUp::isExpired() const {
    return age >= lifetime;
}

File: csharp/03_advanced/SnakeGame/Models/PowerUp.cs
using System;

namespace SnakeGame.Models
{
    public enum PowerUpType
    {
        SpeedBoost,
        Invincibility,
        SlowTime,
        ExtraLife,
        ScoreMultiplier
    }

    public class PowerUp
    {
        public int X { get; set; }
        public int Y { get; set; }
        public PowerUpType Type { get; set; }
        public string Color { get; set; }
        public string Symbol { get; set; }
        public int Duration { get; set; }
        public int Lifetime { get; set; }
        public int Age { get; set; }
        public bool IsExpired => Age >= Lifetime;

        public PowerUp(int x, int y, PowerUpType type)
        {
            X = x;
            Y = y;
            Type = type;
            Age = 0;
            Lifetime = 300; // 5 секунд при 60 FPS

            switch (type)
            {
                case PowerUpType.SpeedBoost:
                    Color = "#00FFFF";
                    Symbol = "⚡";
                    Duration = 300;
                    break;
                case PowerUpType.Invincibility:
                    Color = "#FFFF00";
                    Symbol = "🛡️";
                    Duration = 300;
                    break;
                case PowerUpType.SlowTime:
                    Color = "#FF00FF";
                    Symbol = "⏱️";
                    Duration = 300;
                    break;
                case PowerUpType.ExtraLife:
                    Color = "#FF4444";
                    Symbol = "❤️";
                    Duration = 0;
                    break;
                case PowerUpType.ScoreMultiplier:
                    Color = "#FFAA00";
                    Symbol = "✖️";
                    Duration = 0;
                    break;
            }
        }

        public void Update()
        {
            Age++;
        }
    }
}
